src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs

```diff
@@ -442,19 +441,125 @@ private BoundExpression VisitObjectInitializerMember(BoundObjectInitializerMembe
                         if (!isRhsNestedInitializer)
                         {
                             // Rewrite as simple assignment.
-                            var rewrittenRight = VisitExpression(assignment.Right);
+                            var rewrittenRight = VisitExpression(right);
                             result.Add(MakeStaticAssignmentOperator(assignment.Syntax, rewrittenAccess, rewrittenRight, false, assignment.Type, used: false));
                             return;
                         }
 
                         break;
                     }
 
+                case BoundKind.ImplicitIndexerAccess:
+                    if (temps == null)
+                    {
+                        temps = ArrayBuilder<LocalSymbol>.GetInstance();
+                    }
+
+                    var implicitIndexer = (BoundImplicitIndexerAccess)left;
+
+                    BoundExpression transformedImplicitIndexer;
+                    var argumentType = implicitIndexer.Argument.Type;
+                    if (TypeSymbol.Equals(argumentType, _compilation.GetWellKnownType(WellKnownType.System_Range), TypeCompareKind.ConsiderEverything))
+                    {
+                        transformedImplicitIndexer = VisitRangePatternIndexerAccess(implicitIndexer, temps, result, inInitializer: true);
+
+                        Debug.Assert(transformedImplicitIndexer is BoundCall
+                        {
+                            Arguments: [BoundPassByCopy { Expression: BoundLocal { LocalSymbol: SynthesizedLocal } },
+                                        BoundPassByCopy { Expression: BoundLocal { LocalSymbol: SynthesizedLocal } }]
+                        });
+                    }
+                    else
+                    {
+                        transformedImplicitIndexer = TransformImplicitIndexerAccess(implicitIndexer,
+                            isRegularAssignmentOrRegularCompoundAssignment: true, inInitializer: true, result, temps, isDynamicAssignment: false);
+
+                        Debug.Assert(transformedImplicitIndexer is BoundArrayAccess { Indices: [BoundLocal { LocalSymbol: SynthesizedLocal }] }
+                            or BoundIndexerAccess { Arguments: [BoundLiteral or BoundLocal { LocalSymbol: SynthesizedLocal }] }
+                            or BoundLocal { LocalSymbol: SynthesizedLocal });
+                    }
+
+                    if (transformedImplicitIndexer is BoundIndexerAccess indexerAccess)
+                    {
+                        rewrittenAccess = VisitIndexerAccess(indexerAccess, isLeftOfAssignment: !isRhsNestedInitializer || indexerAccess.Indexer.RefKind != RefKind.None);
```

Comment:
>rewrittenAccess = VisitIndexerAccess(indexerAccess, isLeftOfAssignment: !isRhsNestedInitializer || indexerAccess.Indexer.RefKind != RefKind.None) [](http://example.com/codeflow?start=24&length=145)

It doesn't feel appropriate to visit node which already has its parts lowered. And it look like we need to do this fix-up only because we called ```TransformImplicitIndexerAccess``` which always transforms the node for the ```isLeftOfAssignment``` situation. But this is not always a situation that we have here. Also, it looks like for the ```isLeftOfAssignment``` situation doing this visit is redundant because calling ```MakeStaticAssignmentOperator``` looks sufficient. My recommendation would be to transform the indexer access in the right mode from the start instead.